desmod/pool.py

```diff
@@ -0,0 +1,207 @@
+"""Pool class for modeling a container of resources.
+
+A pool models a container of items or resources. Pool is similar to the :class:
+`simpy.resources.Container`, but with additional events when the Container is
+empty or full. Users can put or get items in the pool with a certain amount as
+a parameter.
+"""
+
+from simpy import Event
+from simpy.core import BoundClass
+
+
+class PoolPutEvent(Event):
+    def __init__(self, pool, amount=1):
+        super(PoolPutEvent, self).__init__(pool.env)
+        self.pool = pool
+        self.amount = amount
+        self.callbacks.append(pool._trigger_get)
+        pool._putters.append(self)
+        pool._trigger_put()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._putters.remove(self)
+            self.callbacks = None
+
+
+class PoolGetEvent(Event):
+    def __init__(self, pool, amount=1):
+        super(PoolGetEvent, self).__init__(pool.env)
+        self.pool = pool
+        self.amount = amount
+        self.callbacks.append(pool._trigger_put)
+        pool._getters.append(self)
+        pool._trigger_get()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._getters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenNewEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenNewEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._new_waiters.append(self)
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._new_waiters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenAnyEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenAnyEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._any_waiters.append(self)
+        pool._trigger_when_any()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._any_waiters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenFullEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenFullEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._full_waiters.append(self)
+        pool._trigger_when_full()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._full_waiters.remove(self)
+            self.callbacks = None
+
+
+class Pool(object):
+    """Simulation pool of arbitrary items.
+
+    `Pool` is similar to :class:`simpy.resources.Container`.
+    It provides a simulation-aware container for managing a pool of objects
```

Comment:
Nit, `simpy.Container` and thus `Pool` do not necessarily model discrete objects. An important use case is to model continuous quantities (i.e. with a `float` level). This doc string should be clarified.